torch/_functorch/aot_autograd.py

```diff
@@ -907,65 +917,72 @@ def is_fun(t):
+        return False
 
 def are_all_mutations_hidden_from_autograd(t):
     if is_traceable_wrapper_subclass(t):
         attrs, _ = t.__tensor_flatten__()
         # If all inner elements are mutations hidden from autograd, then it is a mutation hidden from autograd.
         return all(are_all_mutations_hidden_from_autograd(getattr(t, attr)) for attr in attrs)
-    else:
+    elif isinstance(t, torch.Tensor):
         assert isinstance(t, FunctionalTensor)
         return torch._functionalize_are_all_mutations_hidden_from_autograd(t.elem)
-
-# new_arg and arg here are either:
-# (1) both a FakeTensor
-# (2) both a traceable tensor subclass that holds a FakeTensor
-# Pre-condition: the two args are the "old" and "new" inputs from running functionalization.
-# When we run functionalization and wrap our inputs into FunctionalTensors,
-# we can detect whether or not an input was mutated by checking to see if the inner tensor has changed
-#
-# Normally it would be enough just to check if arg is new_arg, which is normally enough for functionalization
-# to confirm that inputs were not mutated when running the user's model with functionalization on.
-# But when we have subclass inputs, we can't rely on that:
-# `from_fun(to_fun(x)) is x` will return False, because the call to `from_fun` constructs
-# a brand new subclass instance: we are calling __tensor_unflatten__, and going
-# from Subclass(FakeTensor) to Subclass(FunctionalTensor(FakeTensor))
-def was_updated(arg, new_arg):
-    if is_traceable_wrapper_subclass(arg):
-        assert is_traceable_wrapper_subclass(new_arg)
-        attrs, _ = arg.__tensor_flatten__()
-        new_attrs, _ = new_arg.__tensor_flatten__()
-        assert attrs == new_attrs
-        # A tensor subclass was updated if any of its inner elements were updated
-        return any(was_updated(getattr(arg, attr), getattr(new_arg, attr)) for attr in attrs)
     else:
-        return arg is not new_arg
-
-# new_arg and arg here are either:
-# (1) both a FakeTensor
-# (2) both a traceable tensor subclass that holds a FakeTensor
-# Pre-condition: the two args are the "old" and "new" inputs from running functionalization.
-# When we run functionalization and wrap our inputs into FunctionalTensors,
-# we can detect whether or not an input was mutated by checking to see if the inner tensor has changed,
-# but shares storage with the old input
-def was_metadata_updated(arg, new_arg):
-    if is_traceable_wrapper_subclass(arg):
-        assert is_traceable_wrapper_subclass(new_arg)
-        attrs, _ = arg.__tensor_flatten__()
-        new_attrs, _ = new_arg.__tensor_flatten__()
-        assert attrs == new_attrs
+        return False
+
+# f_arg here is either
+# (1) A FunctionalTensor(_to_functional_tensor(FakeTensor))
+# (2) A traceable tensor subclass that holds a FunctionalTensor
+# Assumption: arg promises to be the "original" tensor wrapped by f_arg
+def has_metadata_mutation(f_arg, arg, *, check_storage_mutation: bool):
```

Comment:
So what exactly does `check_storage_mutation` do?

Reply:
I'll rename the var + comment this more clearly.  "storage mutation" is a subset of "metadata mutation". So:

(1) check_storage_mutation=True: only return true if there is a storage mutation
(2) check_storage_mutation=False: return true if there is any metadata mutation (including a storage mutation)